### PR TITLE
[TIR][Meta-Schedule] Tuple-reduction scheduling support

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR improves our TIR scheduling primitives/transformations (rfactor & cross-thread reduction) designed for reduction operators, so that they can be applied to blocks of tuple-reduction.

Most of the code changes are of unit tests.

cc @junrushao1994

### CODE REVIEW

**Comment by junrushao1994** [comment]
Thanks a lot!! Having tuple reduction in TIR is definitely one of my top priority! Need some time to recharge this week and will review next week!

File: tests/python/unittest/test_tir_transform_lower_cross_thread_reduction.py [hunk 1]
```diff
@@ -44,7 +47,7 @@ def loop_split(a: T.handle, b: T.handle) -> None:
             with T.block("B"):
                 vi = T.axis.S(128, i)
                 vk = T.axis.R(128, ko * 32 + ki)
-                T.reads([B[vi], A[vi, vk]])
+                T.reads([A[vi, vk]])
```
Comment (junrushao) [inline tests/python/unittest/test_tir_transform_lower_cross_thread_reduction.py:50]: Why are we removing `B[vi]` from `T.reads`? It's actually read, isn't it?

Comment (MasterJH5574) [reply tests/python/unittest/test_tir_transform_lower_cross_thread_reduction.py:50]: I guess starting from some commit, we removed all reduction buffers from `T.reads` 🤔.

One example in other test cases is https://github.com/apache/tvm/blob/9671aee942503815ad2a586406eef11391287ee5/tests/python/unittest/test_tvmscript_roundtrip.py#L2716-L2717

But I can’t remember when it starts from.

Comment (junrushao) [reply tests/python/unittest/test_tir_transform_lower_cross_thread_reduction.py:50]: I see! I just checked myself and found the same thing :-) Thanks for a heads-up!

Comment (MasterJH5574) [reply tests/python/unittest/test_tir_transform_lower_cross_thread_reduction.py:50]: My pleasure!

File: src/meta_schedule/schedule_rule/cross_thread_reduction.cc [hunk 2]
```diff
@@ -184,6 +184,13 @@ class CrossThreadReductionNode : public ScheduleRuleNode {
    */
   std::tuple<bool, tir::LoopRV, tir::BlockRV, tir::LoopRV> GetComputeTargetLoopAndBlock(
       const tir::Schedule& sch, const tir::BlockRV& block_rv) {
+    // Step 0. Due to technical reason of some primitives (e.g., compute-at), if the block is doing
+    // a tuple reduction, fusion is temporarily not supported.
```
Comment (junrushao) [inline src/meta_schedule/schedule_rule/cross_thread_reduction.cc:188]: nit: let's improve the doc a bit by being more specific, like, "fusion with the subsequent consumer is not allowed"

File: src/tir/schedule/analysis/analysis.cc [hunk 3]
```diff
@@ -1514,115 +1519,235 @@ class PatternMatcher : public ExprVisitor {
...
+5) The body of the innermost LetStmt should be a single BufferStore or a SeqStmt of BufferStores
+6) The number of BufferStores under the block body should equal the number of BufferStores in the block init, and thereby equal the number of LetStmts above
+7) The variables bound by the LetStmts in the block body must all directly serve as values of the BufferStores inside, and the stored values of the BufferStores can only be those variables
+8) The variables stored by the BufferStores in the block body should be all different
+9) The buffers written by the BufferStores in the block body should be all different
+10) The buffers initialized in the block init and written in the block body should match
+11) The buffers written by the block should have same shape
+12) The indices of all BufferStores in the reduction block should be the same)";
+
+void ErrorRFactorCrossThreadReductionNotApplicable(const Optional<ScheduleState>& self, Block block,
```
Comment (junrushao) [inline src/tir/schedule/analysis/analysis.cc:1544]: nit: move the reducer-related logic into a separate file given they are relatively isolated and the logic is much longer

**Review by junrushao — APPROVED** [review]
LGTM!

**Pull request merged**